File: map_lib/test/test_align_svd.py
#!/usr/bin/env python
import os

import numpy as np
from scipy.spatial.transform import Rotation

from ament_index_python.packages import get_package_share_directory

from map_lib.type import Map
from map_lib.association.align import AlignSVD

SHARE_PATH = get_package_share_directory("map_lib")
REFERENCE_PATH = os.path.join(SHARE_PATH, "test", "reference_map.csv")


class TestSVD:
    def test_svd(self):
        reference_map = Map.from_csv(REFERENCE_PATH)

        R = np.array(Rotation.from_euler('z', 35, degrees=True).as_matrix()[:2, :2])
        t = np.array([0.2, 0.1])
        transformed_map = Map.from_csv(REFERENCE_PATH).transform(R, t)

        transform = AlignSVD(
            reference_map.positions(),
            transformed_map.positions()).find_transformation()
        assert np.isclose(transform[:2, :2], R).all()
        assert np.isclose(transform[:2, 2], t).all()

File: map_lib/test/test_align_svd.cpp
#include <gmock/gmock.h>

#include <filesystem>

#include <ament_index_cpp/get_package_share_directory.hpp>
#include <eufs_gmock_matchers/eigen_matchers.hpp>

#include "map_lib/association/align/align_svd.hpp"
#include "map_lib/io/reader.hpp"
#include "map_lib/transform.hpp"
#include "map_lib/type/cone.hpp"
#include "map_lib/type/map.hpp"

using eufs::map::Cone;
using eufs::map::Map;

TEST(CheckICP, NormalRotation) {
  auto package_share_directory = ament_index_cpp::get_package_share_directory("map_lib");
  auto map_path = std::filesystem::path(package_share_directory) / "test" / "reference_map.csv";
  const Map<Cone> map(eufs::map::io::LoadCones<Cone>(map_path));

  Eigen::Vector2d trans{0.2, 0.1};

  const double angle{7 * M_PI / 36};
  const Map<Cone> transformed_map(eufs::map::Transformation::Transform(map, angle, trans));

  const auto alignment =
    eufs::map::association::align::AlignSVD(map.GetPositions(), transformed_map.GetPositions());
  const Eigen::Affine2d res = alignment.FindTransformation();

  ASSERT_THAT(res.translation(), eufs::testing::matchers::EigenEq(trans));
  ASSERT_THAT(res.rotation(), eufs::testing::matchers::EigenEq(
                      Eigen::Rotation2D<double>(angle).toRotationMatrix()));
}
